power_grid_model_c/power_grid_model_cpp/include/power_grid_model_cpp/model.hpp

```diff
@@ -0,0 +1,74 @@
+// SPDX-FileCopyrightText: Contributors to the Power Grid Model project <powergridmodel@lfenergy.org>
+//
+// SPDX-License-Identifier: MPL-2.0
+
+#pragma once
+#ifndef POWER_GRID_MODEL_CPP_MODEL_HPP
+#define POWER_GRID_MODEL_CPP_MODEL_HPP
+
+#include "basics.hpp"
+#include "dataset.hpp"
+#include "handle.hpp"
+#include "options.hpp"
+
+#include "power_grid_model_c/model.h"
+
+namespace power_grid_model_cpp {
+class Model {
+  public:
+    Model(double system_frequency, DatasetConst const& input_dataset)
+        : model_{handle_.call_with(PGM_create_model, system_frequency, input_dataset.get())} {}
+    Model(Model const& other) : model_{handle_.call_with(PGM_copy_model, other.model_.get())} {}
+    Model& operator=(Model const& other) {
+        if (this != &other) {
+            model_.reset(handle_.call_with(PGM_copy_model, other.model_.get()));
+        }
+        return *this;
+    }
+    Model(Model&& other) noexcept : handle_{std::move(other.handle_)}, model_{std::move(other.model_)} {}
+    Model& operator=(Model&& other) noexcept {
+        if (this != &other) {
+            handle_ = std::move(other.handle_);
+            model_ = std::move(other.model_);
+        }
+        return *this;
+    }
+    ~Model() { model_.reset(); }
```

Comment:
Maybe the last comment. Why do we need this? The destructor can be just defaulted?

Reply:
It can, but sonarcloud doesn't like it. It was written this way to follow rule of 5. If we don't care, which I don't think is a problem here I can just `default` it.

Reply:
the fundamental reason is that we have a custom copy constructor (namely the `copy_model` functionality). the Sonar Cloud proposes that if you do one thing yourself, then you probably want to do everything yourself to prevent bugs, even if that "everything" is trivial and the same as the default

Reply:
It may be a good idea to keep there as a remainder if the destructor actually needs to change from a trivial one at some point. Leaving it as `default` could lead to overlooking this later on.

Reply:
I do not understand the logic. Rule of five says you need to declare them all. But it does not say you need to manually define it. `=default` is perfect. We really do not do anything special in the destructor.

Reply:
Now I agree with @TonyXiang8787 , if 4 out of 5 are explicitly defined and the destructor is declared and `default`ed, that should indicate clearly enough what was the intention to the future developer. I will `default` it then.